taichi/program/async_utils.cpp

```diff
@@ -269,9 +308,10 @@ TaskMeta *get_task_meta(IRBank *ir_bank, const TaskLaunchRecord &t) {
   // accesses (e.g., a = x[i + 1]), we don't treat it as completely
   // overwriting the value state (e.g., for i in x: x[i] = 0).
   for (auto &state : meta.output_states) {
-    if (state.type == AsyncState::Type::value) {
-      if (meta.element_wise.find(state.snode) == meta.element_wise.end() ||
-          !meta.element_wise[state.snode]) {
+    if (state.type == AsyncState::Type::value && state.holds_snode()) {
+      const auto *sn = state.snode();
+      if (meta.element_wise.find(sn) == meta.element_wise.end() ||
+          !meta.element_wise[sn]) {
```

Comment:
This should also apply to global temps because we never completely overwriting the state. 

Maybe we should directly write sth like this in the above `gather_statements`, whose correctness is clearer to me:
```C++
if (stmt->is<GlobalTemporaryStmt>()) {
  meta.input_states.insert(AsyncState::for_global_tmp(t.kernel));
  meta.output_states.insert(AsyncState::for_global_tmp(t.kernel));
}
```

This may cause some tasks only reading the global temp buffer not swappable, but I think there won't be >1 consecutive tasks reading the buffer without writing.

Reply:
Marking `GlobalTemporaryStmt` whenever we see it, as both input and output states sounds good to me!

Reply:
> https://github.com/taichi-dev/taichi/pull/2061#discussion_r528710425

Ha, that's actually where I started (locally)... Done